### PR TITLE
.Net: Experimental Feature - Assistants API

### PR DESCRIPTION
**Pull request by crickman**
### Motivation and Context

This is an experimental feature that integrates Open AI assistants into Semantic Kernel:

https://devblogs.microsoft.com/semantic-kernel/assistants-the-future-of-semantic-kernel/

### Description

This change introduces an assistant framework that integrates with semantic-kernel patterns, particularly around function calling.

https://platform.openai.com/docs/api-reference/assistants

Assistant usage is showcased in two `KernelSyntaxExamples`.

This initial version does not support:

- Files
- Retrieval
- Code Interpreter

image

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Review by alliscode — APPROVED** [review]

**Review by markwallace-microsoft — CHANGES_REQUESTED** [review]
@glahaye we already have an Experimental folder. Can you move this stuff here semantic-kernel\dotnet\src\Experimental

**Comment by glahaye** [comment]
> @glahaye we already have an Experimental folder. Can you move this stuff here semantic-kernel\dotnet\src\Experimental

That's where the files already are!

**Comment by markwallace-microsoft** [comment]
> > @glahaye we already have an Experimental folder. Can you move this stuff here semantic-kernel\dotnet\src\Experimental
> 
> That's where the files already are!

Sorry my bad

**Review by markwallace-microsoft — CHANGES_REQUESTED** [review]
The experimental feature should be feature complete and have unit tests before merging to main.

**Comment by crickman** [comment]
> The experimental feature should be feature complete and have unit tests before merging to main.

Currently "DRAFT" PR.

File: dotnet/src/Experimental/Assistants/Extensions/OpenAIRestExtensions.cs
Comment (RogerBarreto) [inline dotnet/src/Experimental/Assistants/Extensions/OpenAIRestExtensions.cs:1]: I understand that this is a core functionality of OpenAIRestContext, is there a strong reason why this was implemented as an extension?

Testing this functionality is "not possible" ATM.

**Comment by crickman** [comment]
> The experimental feature should be feature complete and have unit tests before merging to main.

Screenshot 2023-11-15 082709

File: dotnet/src/Experimental/Assistants/Extensions/OpenAIRestExtensions.cs
Comment (crickman) [reply dotnet/src/Experimental/Assistants/Extensions/OpenAIRestExtensions.cs:1]: I understand the concert and complication with unit-testing "connectect" clients.

I think Gil has used Mock<HttpMessageHandler> to cover this area.

File: .github/_typos.toml [hunk 1]
```diff
@@ -18,6 +18,7 @@ extend-exclude = [
 
 [default.extend-words]
 ACI = "ACI" # Azure Container Instance
+exercize = "exercize" #test typos
```
Comment (markwallace-microsoft) [inline .github/_typos.toml:21]: Can we ignore the misspelling just were needed so we don't end up with this misspelling elsewhere in the code

File: dotnet/src/SemanticKernel.Abstractions/SemanticKernel.Abstractions.csproj [hunk 2]
```diff
@@ -28,5 +28,6 @@
     <InternalsVisibleTo Include="SemanticKernel.Extensions.UnitTests" />
     <InternalsVisibleTo Include="Microsoft.SemanticKernel.Planners.Core.UnitTests" />
     <InternalsVisibleTo Include="DynamicProxyGenAssembly2" /> <!-- Moq -->
+    <InternalsVisibleTo Include="Microsoft.SemanticKernel.Experimental.Assistants" />
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/SemanticKernel.Abstractions.csproj:31]: Nothing to change now but we should revisit adding this to the abstractions and core projects. What is it you need that is internal?

File: dotnet/samples/KernelSyntaxExamples/Program.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ public static async Task Main(string[] args)
         using CancellationTokenSource cancellationTokenSource = new();
         CancellationToken cancelToken = cancellationTokenSource.ConsoleCancellationToken();
 
-        string? defaultFilter = null; // Modify to filter examples
+        string? defaultFilter = "Example71_AssistantDelegation"; // Modify to filter examples
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Program.cs:24]: Revert this change

File: dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+         string resourcePath,
+         params string[] messages)
+    {
+        return ChatAsync(resourcePath, null, messages);
+    }
+
+    private static async Task ChatAsync(
+        string resourcePath,
+        IEnumerable<ISKFunction>? functions,
+        params string[] messages)
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs:133]: Can you change order of parameters make functions default to null and then remove the override

File: dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        {
+            var messageUser = await thread.AddUserMessageAsync(message).ConfigureAwait(true);
+            DisplayMessage(messageUser);
+
+            var assistantMessages = await thread.InvokeAsync(assistant).ConfigureAwait(true);
+            DisplayMessage(assistantMessages);
+        }
+    }
+
+    private static void DisplayMessage(IEnumerable<IChatMessage> messages)
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs:157]: `DisplayMessages`

File: dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        foreach (var message in messages)
+        {
+            DisplayMessage(message);
+        }
+    }
+
+    private static void DisplayMessage(IChatMessage message)
+    {
+        Console.WriteLine($"[{message.Id}]");
+        Console.WriteLine($"# {message.Role}: {message.Content}");
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs:168]: Marge these lines into the previous method

File: dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        Console.WriteLine("======== Run:AsFunction ========");
+
+        var assistant =
+            await AssistantBuilder.FromDefinitionAsync(
+                TestConfiguration.OpenAI.ApiKey,
+                OpenAIFunctionEnabledModel,
+                EmbeddedResource.Read("Assistants.ParrotAssistant.yaml"));
+
+        IKernel bootstraper = new KernelBuilder().Build();
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs:107]: Variable us misspelled but why not use `kernel` as the name for clarity

File: dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        Console.WriteLine("======== Example71_AssistantDelegation ========");
+
+        if (TestConfiguration.OpenAI.ApiKey == null)
+        {
+            Console.WriteLine("OpenAI apiKey not found. Skipping example.");
+            return;
+        }
+
+        IKernel bootstraper = new KernelBuilder().Build();
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs:33]: All other samples use variable name of `kernel`

File: dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+        var functions = bootstraper.ImportFunctions(new MenuPlugin(), nameof(MenuPlugin));
+
+        var assistant1 =
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs:37]: Change name to `toolAssistant`

File: dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            await AssistantBuilder.FromDefinitionAsync(
+                TestConfiguration.OpenAI.ApiKey,
+                model: OpenAIFunctionEnabledModel,
+                template: EmbeddedResource.Read("Assistants.ToolAssistant.yaml"),
+                functions: functions.Values);
+
+        var assistant2 =
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs:44]: Change name to `parrotAssistant`

File: dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            await AssistantBuilder.FromDefinitionAsync(
+                TestConfiguration.OpenAI.ApiKey,
+                model: OpenAIFunctionEnabledModel,
+                template: EmbeddedResource.Read("Assistants.ParrotAssistant.yaml"));
+
+        var helperAssistants = Import(assistant1, assistant2).ToArray();
+
+        var assistant3 =
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs:52]: Change name to `toolAssistant`

File: dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            var messageUser = await thread.AddUserMessageAsync(message).ConfigureAwait(true);
+            DisplayMessage(messageUser);
+
+            var assistantMessages = await thread.InvokeAsync(assistant).ConfigureAwait(true);
+            DisplayMessage(assistantMessages);
+        }
+    }
+
+    private static void DisplayMessage(IEnumerable<IChatMessage> messages)
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example71_AssistantDelegation.cs:90]: Same comments the other sample

File: dotnet/src/Experimental/Assistants/AssistantBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        return
+            await Assistant.CreateAsync(
+                new OpenAIRestContext(this._apiKey!, this._httpClientProvider),
+                new OpenAIChatCompletion(this._model.Model, this._apiKey!),
+                this._model,
+                this._functions.Values,
+                cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Assistants/AssistantBuilder.cs:58]: inherited from?

File: dotnet/src/Experimental/Assistants/AssistantBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    public AssistantBuilder WithOpenAIChatCompletionService(string model, string apiKey)
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Assistants/AssistantBuilder.cs:59]: Will this be passed to the Kernel? If yes why not just do `WithKernel`?

For `Kernel` we are looking to remove the Builder pattern in favour of just doing `new Kernel(...)`

File: dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs (hunk 4 above)
Comment (crickman) [reply dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs:133]: Doesn't params need to be the last parameter?

File: dotnet/src/Experimental/Assistants/AssistantBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    /// <inheritdoc/>
+    public AssistantBuilder WithFunction(ISKFunction tool)
+    {
+        this._functions[tool.GetQualifiedName()] = tool;
+
+        return this;
+    }
+
+    /// <inheritdoc/>
+    public AssistantBuilder WithFunctions(IEnumerable<ISKFunction> tools)
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Assistants/AssistantBuilder.cs:127]: We will be changing from function collections to plugin collections in the Kernel soon. Another reason why `WithKernel` would be a better long term strategy.

File: dotnet/src/Experimental/Assistants/AssistantResponse.cs [hunk 16]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel.Experimental.Assistants;
+
+/// <summary>
+/// Response from assistant when called as a <see cref="ISKFunction"/>.
+/// </summary>
+public class AssistantResponse
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Assistants/AssistantResponse.cs:10]: Should this be public? Not a big deal now as this is experimental but one to watch later.

File: dotnet/src/Experimental/Assistants/IAssistant.cs [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.SemanticKernel.Experimental.Assistants;
+
+/// <summary>
+/// Represents an assistant that can call the model and use tools.
+/// </summary>
+public interface IAssistant
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Assistants/IAssistant.cs:12]: Do we expect customers to implement their own Assistants?

We are favouring public classes because once we ship and API we cannot change it. Not an issue now but requires thought later.

File: dotnet/src/Experimental/Assistants/IChatMessage.cs [hunk 18]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.ObjectModel;
+
+namespace Microsoft.SemanticKernel.Experimental.Assistants;
+
+/// <summary>
+/// Represents a message that is part of an assistant thread.
+/// </summary>
+public interface IChatMessage
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Assistants/IChatMessage.cs:10]: We already have this abstraction in core, can we extend that?

File: dotnet/src/Experimental/Assistants/IChatThread.cs [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.SemanticKernel.Experimental.Assistants;
+
+/// <summary>
+/// Represents a thread that contains messages.
+/// </summary>
+public interface IChatThread
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Assistants/IChatThread.cs:12]: Do we expect customers to implement this?

File: dotnet/src/Experimental/Assistants/AssistantResponse.cs (hunk 16 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/AssistantResponse.cs:10]: I believe it might need to be consumed by callers for certain scenarios, yes.  That's why I moved it out of the models folder.  Agree this needs to be watched.

File: dotnet/src/Experimental/Assistants/AssistantBuilder.cs (hunk 15 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/AssistantBuilder.cs:127]: Yes, I plan to align w/ plugin pattern _as soon as available_

Comment (crickman) [reply dotnet/src/Experimental/Assistants/AssistantBuilder.cs:127]: I don't believe we always want every single function that might be defined in a kernel...the choice should be up to the user.

File: dotnet/src/Experimental/Assistants/AssistantBuilder.cs (hunk 14 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/AssistantBuilder.cs:59]: This was a joint decision w/ Matthew...we are certainly free to layer on different builder methods.   Generally, assistant doesn't need just any kernel instance...it only needs chatcompletion and an api key.  Besides I thought our recommendation was not to utilize just one kernel...but yeah...make a lot!  I like the clean ownership between Assistant and its kernel.

File: dotnet/src/Experimental/Assistants/IAssistant.cs (hunk 17 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/IAssistant.cs:12]: This allows for mocking in the callers application pattern.

File: dotnet/src/Experimental/Assistants/IChatMessage.cs (hunk 18 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/IChatMessage.cs:10]: I'm not sure the binding has more value than headache.  Can we take that as a follow-up please?  (Making a list).

File: dotnet/src/Experimental/Assistants/IChatThread.cs (hunk 19 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/IChatThread.cs:12]: This allows for mocking in the callers application pattern.

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs (hunk 4 above)
Comment (crickman) [reply dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs:133]: Discussed in teams

File: .github/_typos.toml (hunk 1 above)
Comment (crickman) [reply .github/_typos.toml:21]: Discussed in teams

File: dotnet/src/Experimental/Assistants/AssistantBuilder.cs (hunk 13 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/AssistantBuilder.cs:58]: Oh shoot....comments...got it.  Yeah, there used to be an interface that I wiped (not necessary...restoring now)

File: dotnet/src/SemanticKernel.Abstractions/SemanticKernel.Abstractions.csproj (hunk 2 above)
Comment (crickman) [reply dotnet/src/SemanticKernel.Abstractions/SemanticKernel.Abstractions.csproj:31]: I think it was testing related...agree to revisist

File: dotnet/src/Experimental/Assistants/AssistantBuilder.cs (hunk 13 above)
Comment (crickman) [reply dotnet/src/Experimental/Assistants/AssistantBuilder.cs]: Brought the comments back

**Comment by crickman** [comment]
image

File: dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs (hunk 6 above)
Comment (crickman) [reply dotnet/samples/KernelSyntaxExamples/Example70_Assistant.cs:168]: This is actually called on line 150

**Review by alliscode — APPROVED** [review]
